source/matrix_free/vector_data_exchange.cc

```diff
@@ -316,6 +318,965 @@ namespace internal
+                }
+              recv_ptr.push_back(recv_indices.size());
+            }
+
+          recv_sm_ptr = recv_ptr;
+          recv_sm_ptr.shrink_to_fit();
+          recv_sm_indices = recv_indices;
+          recv_sm_indices.shrink_to_fit();
+          recv_sm_len = recv_len;
+          recv_sm_len.shrink_to_fit();
+        }
+
+        std::vector<unsigned int>
+        procs_of_sm(const MPI_Comm &comm, const MPI_Comm &comm_shared)
+        {
+          std::vector<unsigned int> ranks_shared;
+
+#ifndef DEAL_II_WITH_MPI
+          Assert(false, ExcNeedsMPI());
+          (void)comm;
+          (void)comm_shared;
+#else
+          // extract information from comm
+          int rank_;
+          MPI_Comm_rank(comm, &rank_);
+
+          const unsigned int rank = rank_;
+
+          // extract information from sm-comm
+          int size_shared;
+          MPI_Comm_size(comm_shared, &size_shared);
+
+          // gather ranks
+          ranks_shared.resize(size_shared);
+          MPI_Allgather(&rank,
+                        1,
+                        MPI_UNSIGNED,
+                        ranks_shared.data(),
+                        1,
+                        MPI_INT,
+                        comm_shared);
+#endif
+
+          return ranks_shared;
+        }
+
+      } // namespace internal
+
+
+
+      Full::Full(
+        const std::shared_ptr<const Utilities::MPI::Partitioner> &partitioner,
+        const MPI_Comm &communicator_sm)
+        : recv_sm_ptr{0}
+        , send_remote_ptr{0}
+        , send_remote_offset{0}
+        , send_sm_ptr{0}
+      {
+        const IndexSet &is_locally_owned = partitioner->locally_owned_range();
+        const IndexSet &is_locally_ghost = partitioner->ghost_indices();
+        const MPI_Comm &communicator     = partitioner->get_mpi_communicator();
+        const std::vector<std::pair<unsigned int, unsigned int>>
+          ghost_indices_within_larger_ghost_set =
+            partitioner->ghost_indices_within_larger_ghost_set();
+
+        this->comm    = communicator;
+        this->comm_sm = communicator_sm;
+
+#ifndef DEAL_II_WITH_MPI
+        Assert(false, ExcNeedsMPI());
+
+        (void)is_locally_owned;
+        (void)is_locally_ghost;
+        (void)ghost_indices_within_larger_ghost_set;
+#else
+        this->n_local_elements  = is_locally_owned.n_elements();
+        this->n_ghost_elements  = is_locally_ghost.n_elements();
+        this->n_global_elements = is_locally_owned.size();
+
+        if (Utilities::MPI::job_supports_mpi() == false)
+          return; // nothing to do in serial case
+
+        std::vector<unsigned int> sm_ranks(
+          Utilities::MPI::n_mpi_processes(comm_sm));
+
+        const unsigned int rank = Utilities::MPI::this_mpi_process(comm);
+
+        MPI_Allgather(
+          &rank, 1, MPI_UNSIGNED, sm_ranks.data(), 1, MPI_UNSIGNED, comm_sm);
+
+        std::vector<unsigned int> owning_ranks_of_ghosts(
+          is_locally_ghost.n_elements());
```

Comment:
Do you want to initialize this to `numbers::invalid_unsigned_int` to be explicit that this is filled in the consensus algorithm below?